pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,381 @@
+			if stringMatch(service, rule.Services) {
+				// Generate the policy if the service is matched to the services specified in ServiceRole.
+				log.Debugf("role %v matched by AccessRule %d", role.Name, i)
+				if policy == nil {
+					policy = &policyproto.Policy{
+						Permissions: []*policyproto.Permission{},
+						Principals:  principals,
+					}
+				}
+				policy.Permissions = append(policy.Permissions, convertToPermission(rule))
+			}
+		}
+
+		if policy != nil {
+			log.Debugf("role %v generated policy %v", role.Name, *policy)
+			rbac.Policies[role.Name] = policy
+		}
+	}
+
+	return &rbacconfig.RBAC{Rules: rbac}
+}
+
+// convertToPermission converts a single AccessRule to a Permission.
+func convertToPermission(rule *rbacproto.AccessRule) *policyproto.Permission {
+	rules := &policyproto.Permission_AndRules{
+		AndRules: &policyproto.Permission_Set{
+			Rules: make([]*policyproto.Permission, 0),
+		},
+	}
+
+	if len(rule.Methods) > 0 {
+		methodRule := permissionForKeyValues(":method", rule.Methods)
+		if methodRule != nil {
+			rules.AndRules.Rules = append(rules.AndRules.Rules, methodRule)
+		}
+	}
+
+	if len(rule.Paths) > 0 {
+		pathRule := permissionForKeyValues(":path", rule.Paths)
+		if pathRule != nil {
+			rules.AndRules.Rules = append(rules.AndRules.Rules, pathRule)
+		}
+	}
+
+	if len(rule.Constraints) > 0 {
+		// Constraint rule is matched with AND semantics, it's invalid if 2 constraints have the same
+		// key and this should already be caught in validation stage.
+		for _, constraint := range rule.Constraints {
+			p := permissionForKeyValues(constraint.Key, constraint.Values)
+			if p != nil {
+				rules.AndRules.Rules = append(rules.AndRules.Rules, p)
+			}
+		}
+	}
+
+	return &policyproto.Permission{Rule: rules}
+}
+
+// convertToPrincipals converts a list of subjects to principals.
+func convertToPrincipals(bindings []*rbacproto.ServiceRoleBinding) []*policyproto.Principal {
+	principals := make([]*policyproto.Principal, 0)
+	for _, binding := range bindings {
+		for _, subject := range binding.Subjects {
+			principals = append(principals, convertToPrincipal(subject))
+		}
+	}
+	return principals
+}
+
+// convertToPrincipal converts a single subject to principal.
+func convertToPrincipal(subject *rbacproto.Subject) *policyproto.Principal {
+	ids := &policyproto.Principal_AndIds{
+		AndIds: &policyproto.Principal_Set{
+			Ids: make([]*policyproto.Principal, 0),
+		},
+	}
+
+	if subject.User != "" {
+		id := principalForKeyValue(userHeader, subject.User)
+		if id != nil {
+			ids.AndIds.Ids = append(ids.AndIds.Ids, id)
+		}
+	}
+
+	if subject.Group != "" {
+		log.Errorf("ignored Subject.group %s, not implemented", subject.Group)
+	}
+
+	if len(subject.Properties) != 0 {
+		// Use a separate key list to make sure the map iteration order is stable, so that the generated
+		// config is stable.
+		var keys []string
+		for k := range subject.Properties {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+
```

Comment:
Why do you need to "sort"?

Reply:
This allow us to iterate the map in a determined order. The go map doesn't guarantee the order of the iteration, so the function might return a different config even if there is no policy change at all. This is bad for outer cache.

Reply:
Thanks for explanation. Shall we also sort keys for Constraints in permission?

Reply:
The constraints in permission is different, it's actually a repeated message that includes a key/value field. So we don't need to do this for it.